File: python/examples/conntest.py
#!/usr/bin/env python3
"""conntest — v0.7.0 connected-messaging round-trip validator.

Output format is byte-identical (modulo dt= timing values) to
c/examples/conntest.c + go/cmd/conntest.  The cross-language gate
is the three diffs.

SPDX-License-Identifier: MIT
"""
import argparse
import bisect
import sys
import time

import bpclient


IDENTITY_REQ = bytes([0x01, 0x02, 0x20, 0x01, 0x24, 0x01])


def now_ms() -> float:
    return time.monotonic() * 1000.0


def percentile(xs: list[float], p: float) -> float:
    if not xs:
        return 0.0
    s = sorted(xs)
    idx = p * (len(s) - 1)
    lo = int(idx)
    hi = lo + 1 if lo + 1 < len(s) else len(s) - 1
    frac = idx - lo
    return s[lo] * (1.0 - frac) + s[hi] * frac


def validate_identity(resp: bytes) -> bool:
    return len(resp) >= 4 and resp[0] == 0x81 and resp[2] == 0x00


def print_identity(resp: bytes) -> None:
    if len(resp) < 14:
        print("[conntest] identity: (body too short)")
        return
    off = 4 + resp[3] * 2
    body = resp[off:]
    if len(body) < 14:
        print("[conntest] identity: (body too short)")
        return
    vendor = body[0] | (body[1] << 8)
    dev = body[2] | (body[3] << 8)
    prod = body[4] | (body[5] << 8)
    major = body[6]
    minor = body[7]
    name_len = body[14] if len(body) > 14 else 0
    if 15 + name_len > len(body):
        name_len = len(body) - 15
    name = body[15:15 + name_len].decode("ascii", "replace")
    print(f"[conntest] identity: Vendor=0x{vendor:04x} DevType=0x{dev:04x} "
          f"Product=0x{prod:04x} fw={major}.{minor} Name='{name}'")


def run_bench(client: bpclient.Client, slot: int, n: int) -> int:
    print(f"[conntest] benchmark: {n} Identity round-trips per transport, slot={slot}")

    ucmm_dt: list[float] = []
    class3_dt: list[float] = []

    # UCMM loop.
    for i in range(n):
        msg = bpclient.Message(
            slot=slot, cip_request=IDENTITY_REQ,
            resp_data=b"", resp_capacity=256, timeout_ms=5000,
        )
        t0 = now_ms()
        client.message_send(msg)
        ucmm_dt.append(now_ms() - t0)
        if not validate_identity(msg.resp_data):
            sys.stderr.write(f"[conntest] UCMM bench: req[{i}] reply invalid\n")
            return 1

    # Class-3 loop.
    spec = bpclient.ConnSpec(
        app_handle=2, encoded_path=bytes([0x01, slot]), path_size=2,
    )
    client.txrx_open(spec)
    try:
        for i in range(n):
            t0 = now_ms()
            resp = client.txrx_msg(spec, IDENTITY_REQ, 256)
            class3_dt.append(now_ms() - t0)
            if not validate_identity(resp):
                sys.stderr.write(f"[conntest] Class3 bench: req[{i}] reply invalid\n")
                return 1
    finally:
        client.txrx_close(spec)

    print(f"[conntest]   UCMM     median dt={percentile(ucmm_dt, 0.50):.2f}ms  "
          f"p95 dt={percentile(ucmm_dt, 0.95):.2f}ms")
    print(f"[conntest]   Class3   median dt={percentile(class3_dt, 0.50):.2f}ms  "
          f"p95 dt={percentile(class3_dt, 0.95):.2f}ms")
    return 0


def main() -> int:
    ap = argparse.ArgumentParser(prog="conntest")
    ap.add_argument("--slot", type=int, default=2,
                    help="backplane slot number (default 2 = L85)")
    ap.add_argument("--n", type=int, default=10,
                    help="number of Identity round-trips on the connection")
    ap.add_argument("--bench", action="store_true",
                    help="run UCMM vs Class3 latency micro-benchmark (100 each)")
    args = ap.parse_args()

    try:
        client = bpclient.Client()
        client.open()
    except Exception:
        sys.stderr.write("[conntest] open failed\n")
        return 2
    try:
        client.open_session()
    except Exception:
        pass

    spec = bpclient.ConnSpec(
        app_handle=1,
        encoded_path=bytes([0x01, args.slot]),
        path_size=2,
    )

    print(f"[conntest] slot={args.slot} N={args.n} app_handle={spec.app_handle}")

    try:
        conn_id, conn_serial = client.txrx_open(spec)
    except Exception as e:
        sys.stderr.write(f"[conntest] txrx_open: {e}\n")
        client.close()
        return 1
    print(f"[conntest] txrx_open  conn_id=0x{conn_id:04x}  serial=0x{conn_serial:04x}")

    dts: list[float] = []
    last_resp = b""
    success = 0
    for i in range(args.n):
        t0 = now_ms()
        try:
            resp = client.txrx_msg(spec, IDENTITY_REQ, 256)
        except Exception:
            resp = b""
        dts.append(now_ms() - t0)
        status = resp[2] if len(resp) >= 3 else 0xFF
        vendor = 0xFFFF
        if validate_identity(resp):
            off = 4 + resp[3] * 2
            if off + 2 <= len(resp):
                vendor = resp[off] | (resp[off + 1] << 8)
        print(f"[conntest] req[{i}] dt={dts[-1]:.2f}ms "
              f"status=0x{status:02x} vendor=0x{vendor:04x}")
        if validate_identity(resp):
            success += 1
            last_resp = resp

    if last_resp:
        print_identity(last_resp)

    close_ok = True
    try:
        client.txrx_close(spec)
    except Exception:
        close_ok = False
    print(f"[conntest] txrx_close {'ok' if close_ok else 'FAIL'}")

    print(f"[conntest] SUMMARY {success}/{args.n} success  "
          f"median dt={percentile(dts, 0.50):.2f}ms")
    pass_ = success == args.n and close_ok
    print(f"[conntest] {'PASS' if pass_ else 'FAIL'}")

    bench_rc = 0
    if args.bench:
        bench_rc = run_bench(client, args.slot, 100)

    client.close()
    return 0 if (pass_ and bench_rc == 0) else 1


if __name__ == "__main__":
    sys.exit(main())

File: c/examples/conntest.c
/*
 * conntest.c — v0.7.0 connected-messaging round-trip validator.
 *
 * Opens a class-3 connection to a slot (default 2 = L85), sends N
 * (default 10) Identity Get_Attributes_All requests over the
 * connection, closes, prints a PASS/FAIL summary.  Output format
 * is intentionally byte-identical (modulo dt= timing values) to
 * the Go + Python conntest CLIs — the cross-language gate is the
 * three diffs.
 *
 * --bench flag adds a UCMM-vs-class3 micro-benchmark: 100 Identity
 * round-trips each, prints median + p95 per transport.  Useful as
 * a v0.8 regression baseline once mbox 0x204 lands.
 *
 * SPDX-License-Identifier: MIT
 */
#include <getopt.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#include "bpclient.h"

static double now_ms(void) {
    struct timespec ts;
    clock_gettime(CLOCK_MONOTONIC, &ts);
    return ts.tv_sec * 1000.0 + ts.tv_nsec / 1.0e6;
}

static int cmp_double(const void *a, const void *b) {
    double da = *(const double *)a, db = *(const double *)b;
    if (da < db) return -1;
    if (da > db) return  1;
    return 0;
}

static double percentile(double *xs, int n, double p) {
    /* In-place sort.  p in [0..1]. */
    if (n <= 0) return 0.0;
    qsort(xs, n, sizeof(double), cmp_double);
    double idx = p * (n - 1);
    int lo = (int)idx;
    int hi = lo + 1 >= n ? n - 1 : lo + 1;
    double frac = idx - lo;
    return xs[lo] * (1.0 - frac) + xs[hi] * frac;
}

/* Identity Get_Attributes_All on class 1 instance 1. */
static const uint8_t identity_req[] = { 0x01, 0x02, 0x20, 0x01, 0x24, 0x01 };

/* Returns 1 on a clean Identity reply (svc 0x81, gs 0x00). */
static int validate_identity_reply(const uint8_t *resp, uint16_t len) {
    return (len >= 4 && resp[0] == 0x81 && resp[2] == 0x00);
}

static void print_identity(const uint8_t *resp, uint16_t len) {
    if (len < 14) {
        printf("[conntest] identity: (body too short)\n");
        return;
    }
    const uint8_t *body = resp + 4 + resp[3] * 2;
    uint16_t blen = len - (uint16_t)(4 + resp[3] * 2);
    if (blen < 14) {
        printf("[conntest] identity: (body too short)\n");
        return;
    }
    uint16_t vendor = body[0] | (body[1] << 8);
    uint16_t dev    = body[2] | (body[3] << 8);
    uint16_t prod   = body[4] | (body[5] << 8);
    uint8_t  major  = body[6];
    uint8_t  minor  = body[7];
    uint8_t  name_len = blen > 14 ? body[14] : 0;
    if (15u + name_len > blen) name_len = (uint8_t)(blen - 15);
    printf("[conntest] identity: Vendor=0x%04x DevType=0x%04x Product=0x%04x "
           "fw=%u.%u Name='%.*s'\n",
           vendor, dev, prod, major, minor, (int)name_len, body + 15);
}

static int run_bench(bp_client_t *cl, int slot, int n) {
    printf("[conntest] benchmark: %d Identity round-trips per transport, slot=%d\n",
           n, slot);

    double *ucmm_dt   = calloc((size_t)n, sizeof(double));
    double *class3_dt = calloc((size_t)n, sizeof(double));
    if (!ucmm_dt || !class3_dt) {
        free(ucmm_dt); free(class3_dt);
        fprintf(stderr, "[conntest] bench: alloc failed\n");
        return 1;
    }

    /* UCMM loop. */
    for (int i = 0; i < n; i++) {
        uint8_t resp[256];
        bp_message_t msg = {
            .slot = (uint8_t)slot,
            .cip_request = identity_req,
            .req_size = sizeof(identity_req),
            .timeout_ms = 5000,
            .resp_data = resp,
            .resp_capacity = sizeof(resp),
        };
        double t0 = now_ms();
        int rc = bp_client_message_send(cl, &msg);
        ucmm_dt[i] = now_ms() - t0;
        if (rc != BP_OK || !validate_identity_reply(resp, msg.resp_len)) {
            fprintf(stderr, "[conntest] UCMM bench: req[%d] failed (rc=%d)\n", i, rc);
            free(ucmm_dt); free(class3_dt);
            return 1;
        }
    }

    /* Class-3 loop — open once, msg N times, close once. */
    uint8_t epath[] = { 0x01, (uint8_t)slot };
    bp_conn_spec_t spec = {
        .app_handle   = 2,
        .options      = 0,
        .encoded_path = epath,
        .path_size    = 2,
        .conn_params  = 0,
    };
    int rc = bp_client_txrx_open(cl, &spec, NULL, NULL);
    if (rc != BP_OK) {
        fprintf(stderr, "[conntest] bench: txrx_open rc=%d (%s)\n", rc, bp_strerror(rc));
        free(ucmm_dt); free(class3_dt);
        return 1;
    }
    for (int i = 0; i < n; i++) {
        uint8_t resp[256];
        uint16_t got = 0;
        double t0 = now_ms();
        int mrc = bp_client_txrx_msg(cl, &spec, identity_req, sizeof(identity_req),
                                      resp, sizeof(resp), &got);
        class3_dt[i] = now_ms() - t0;
        if (mrc != BP_OK || !validate_identity_reply(resp, got)) {
            fprintf(stderr, "[conntest] Class3 bench: req[%d] failed (rc=%d)\n", i, mrc);
            (void)bp_client_txrx_close(cl, &spec);
            free(ucmm_dt); free(class3_dt);
            return 1;
        }
    }
    (void)bp_client_txrx_close(cl, &spec);

    double um = percentile(ucmm_dt,   n, 0.50);
    double up = percentile(ucmm_dt,   n, 0.95);
    double cm = percentile(class3_dt, n, 0.50);
    double cp = percentile(class3_dt, n, 0.95);
    printf("[conntest]   UCMM     median dt=%.2fms  p95 dt=%.2fms\n", um, up);
    printf("[conntest]   Class3   median dt=%.2fms  p95 dt=%.2fms\n", cm, cp);

    free(ucmm_dt); free(class3_dt);
    return 0;
}

int main(int argc, char *argv[]) {
    int slot = 2;
    int N = 10;
    int bench = 0;

    static struct option opts[] = {
        {"slot",  required_argument, 0, 's'},
        {"n",     required_argument, 0, 'n'},
        {"bench", no_argument,       0, 'b'},
        {0,0,0,0}
    };
    int oc, idx;
    while ((oc = getopt_long(argc, argv, "s:n:b", opts, &idx)) != -1) {
        if      (oc == 's') slot  = (int)strtol(optarg, NULL, 0);
        else if (oc == 'n') N     = (int)strtol(optarg, NULL, 0);
        else if (oc == 'b') bench = 1;
    }

    bp_client_t *cl = NULL;
    if (bp_client_open(&cl) != BP_OK) {
        fprintf(stderr, "[conntest] open failed\n");
        return 2;
    }
    bp_client_open_session(cl, NULL);

    uint8_t epath[] = { 0x01, (uint8_t)slot };
    bp_conn_spec_t spec = {
        .app_handle   = 1,
        .options      = 0,
        .encoded_path = epath,
        .path_size    = 2,
        .conn_params  = 0,
    };

    printf("[conntest] slot=%d N=%d app_handle=%u\n", slot, N, spec.app_handle);

    uint16_t conn_id = 0, conn_serial = 0;
    int rc = bp_client_txrx_open(cl, &spec, &conn_id, &conn_serial);
    if (rc != BP_OK) {
        fprintf(stderr, "[conntest] txrx_open rc=%d (%s)\n", rc, bp_strerror(rc));
        bp_client_close(cl);
        return 1;
    }
    printf("[conntest] txrx_open  conn_id=0x%04x  serial=0x%04x\n", conn_id, conn_serial);

    double *dts = calloc((size_t)N, sizeof(double));
    if (!dts) {
        fprintf(stderr, "[conntest] alloc failed\n");
        (void)bp_client_txrx_close(cl, &spec);
        bp_client_close(cl);
        return 2;
    }

    uint8_t last_resp[256];
    uint16_t last_len = 0;
    int success = 0;
    for (int i = 0; i < N; i++) {
        uint8_t resp[256];
        uint16_t got = 0;
        double t0 = now_ms();
        int mrc = bp_client_txrx_msg(cl, &spec, identity_req, sizeof(identity_req),
                                      resp, sizeof(resp), &got);
        dts[i] = now_ms() - t0;
        uint8_t st = (got >= 3) ? resp[2] : 0xFF;
        uint16_t vendor = (got >= 6 && validate_identity_reply(resp, got))
                          ? (resp[4 + resp[3] * 2] | (resp[4 + resp[3] * 2 + 1] << 8))
                          : 0xFFFF;
        printf("[conntest] req[%d] dt=%.2fms status=0x%02x vendor=0x%04x\n",
               i, dts[i], st, vendor);
        if (mrc == BP_OK && validate_identity_reply(resp, got)) {
            success++;
            memcpy(last_resp, resp, got);
            last_len = got;
        }
    }

    if (last_len > 0) print_identity(last_resp, last_len);

    int crc = bp_client_txrx_close(cl, &spec);
    printf("[conntest] txrx_close %s\n", crc == BP_OK ? "ok" : "FAIL");

    double median = percentile(dts, N, 0.50);
    printf("[conntest] SUMMARY %d/%d success  median dt=%.2fms\n", success, N, median);
    int pass = (success == N) && (crc == BP_OK);
    printf("[conntest] %s\n", pass ? "PASS" : "FAIL");

    free(dts);

    int bench_rc = 0;
    if (bench) bench_rc = run_bench(cl, slot, 100);

    bp_client_close(cl);
    return (pass && bench_rc == 0) ? 0 : 1;
}
